NuRadioMC/SignalProp/analyticraytracing.py

```diff
@@ -1718,11 +1750,448 @@ def get_path(self, iS, n_points=1000):
+        c = speed_of_light * units.m / units.ns
+
+        t_0 = []
+        t_1 = []
+
+        for i in range(acc-1):
+
+            if path_type == 0:
+
+                n1 = ice_n.get_birefringence_index_of_refraction(p[0, i])[0]
+                n2 = ice_n.get_birefringence_index_of_refraction(p[0, i])[1]
+                n3 = ice_n.get_birefringence_index_of_refraction(p[0, i])[2]
+
+                dx = p[0, i+1, 0] - p[0, i, 0]
+                dy = p[0, i+1, 1] - p[0, i, 1]
+                dz = p[0, i+1, 2] - p[0, i, 2]
+
+            elif path_type == 1:
+
+                n1 = ice_n.get_birefringence_index_of_refraction(p[1, i])[0]
+                n2 = ice_n.get_birefringence_index_of_refraction(p[1, i])[1]
+                n3 = ice_n.get_birefringence_index_of_refraction(p[1, i])[2]
+
+                dx = p[1, i+1, 0] - p[1, i, 0]
+                dy = p[1, i+1, 1] - p[1, i, 1]
+                dz = p[1, i+1, 2] - p[1, i, 2]
+
+            else:
+                self.__logger.error('error: wrong path type')
+
+            direction = np.array([dx, dy, dz])
+            len_diff = np.linalg.norm(direction)
+            direction = direction / len_diff
+
+            s = self.get_effective_index_fast(n1, n2, n3, direction)
+
+            n_0 = s[0]
+            n_1 = s[1]
+
+            t_0.append(len_diff * n_0 / c)
+            t_1.append(len_diff * n_1 / c)
+
+            l = open("ana_l.txt","a+")
+            l.write(str(len_diff) + "\n")
+            l.close()
+
+        Dt = np.sum(t_0) - np.sum(t_1)
+
+        return(Dt, np.array(t_0), np.array(t_1))
+
+    def get_path_polarization(self, source, antenna, path_type=0, acc=1000):
+        
+        """
+        Function for the polarization of the two states of the wave along its path and the path length
+        
+        Parameters
+        ------------
+        source: numpy.array
+            position of the interaction
+        antenna: numpy.array
+            position of the receiving antenna
+        path_type: int
+            refers to the short or long path of the radio wave (0 for short, 1 for long)
+        acc: int
+            step number of the ray tracer
+        
+        Returns
+        ------------
+        solution_type: list:    [sky_e, sky_o, l_path, N0, N1]
+        meaning:                [0] - sky_0 (numpy.array) - polarization vectors of the ordinary ray along the trace
+                                [1] - sky_1 (numpy.array) - polarization vectors of the extraordinary ray along the trace
+                                [2] - l_path (numpy.array) - length of the path increment
+                                [3] - N0 (numpy.array) - effective refractive index vectors of the ordinary ray along the trace
+                                [4] - N1 (numpy.array) - effective refractive index vectors of the extraordinary ray along the trace
+        """
+        
+        ice_n = self.__medium
+        p = self.get_3d_trace(source, antenna, acc)
```

Comment:
you already called this function in the `get_time_delays` function. Hence, you are repeating the ray tracing twice. You don't need to do that. The ray tracing was anyway done already. There is also no need of passing `source` and `antenna` to the function. This information already exists as member variables `_X1` and `_X2`. You can assume that the user already called the function `set_start_and_end_point(source, antenna)`

Reply:
Done.